template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py

```diff
@@ -1,23 +1,49 @@
 from functools import partial
+import numpy as np
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
+from workflow.torch import ModuleCompose, model_device
 
-from workflow.torch import ModuleCompose
+from {{cookiecutter.package_name}} import problem, architecture
 
 
-def SimpleModel():
-    return ModuleCompose(
-        nn.Conv2d(1, 32, 3, 1),
-        F.relu,
-        nn.Conv2d(32, 64, 3, 1),
-        F.relu,
-        partial(F.max_pool2d, kernel_size=2),
-        nn.Dropout2d(0.25),
-        partial(torch.flatten, start_dim=1),
-        nn.Linear(9216, 128),
-        F.relu,
-        nn.Dropout2d(0.5),
-        nn.Linear(128, 10),
-        partial(F.log_softmax, dim=1),
-    )
+class SimpleModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.core = ModuleCompose(
+            nn.Conv2d(1, 32, 3, 1),
+            F.relu,
+            nn.Conv2d(32, 64, 3, 1),
+            F.relu,
+            partial(F.max_pool2d, kernel_size=2),
+            nn.Dropout2d(0.25),
+            partial(torch.flatten, start_dim=1),
+            nn.Linear(9216, 128),
+            F.relu,
+            nn.Dropout2d(0.5),
+            nn.Linear(128, 10),
+            partial(F.log_softmax, dim=1),
+        )
+
+    def preprocessed(self, images):
```

Comment:
I'm not sure about having this as a class method, it's quite general. Maybe just a function inside `package.architecture`?

Reply:
It is only meant to be used by `model` so I think it should be inside model.py

The reason I put it as a method is to allow adversarial examples easily in `train_batch` as we discussed yesterday.

Something like this
```
def train_batch(...):
    preprocessed = model.preprocessed(images)
    prediction = model(adversarial_image(preprocessed))
```

Reply:
Ah, you're right about that. Perhaps if we note that we reuse the same code in multiple models it could be in a separate function that the models import and use in their own `self.preprocess`.